lark/tree_templates.py

```diff
@@ -55,7 +55,7 @@ def _get_tree(self, template: TreeOrCode) -> Tree[str]:
     def __call__(self, template: Tree[str]) -> 'Template':
         return Template(template, conf=self)
 
-    def _match_tree_template(self, template: TreeOrCode, tree: TreeOrCode) -> Optional[Dict[str, TreeOrCode]]:
+    def _match_tree_template(self, template: TreeOrCode, tree: Tree) -> Optional[Dict[str, Tree]]:

```

Comment:
The annotation for the argument `tree` isn't quite correct based on the rest of the function, or the test `template == tree` in the `if isinstance(template, str)` case below is not needed.

Reply:
Actually, the problem wasn't with these functions returning `TreeOrCode` (although it's a bit of a misnomer), it's the problem of everything else not accepting `TreeOrToken` as an input in the type signature.

Reply:
@henryiii I think the only thing needed it to change back `tree: Tree` into `tree: TreeOrCode`

Reply:
Then it's taking in `Tree | str` and returning `Tree` without converting `str -> Tree`. So MyPy (rightly) complains.

Reply:
`match` and `search` still take `TreeOrCode`. Just not `_match_tree_template`.

Reply:
The return type should also be `Tree | str`, that was correct before. Wherever else that causes problems when receiving arguments need to be changed to accept `Tree | str`. For example, the template `Tree('data', ["$var"])` applied to `Tree('data', ["value"])` should output `{'$var': "value"}`.

Reply:
Okay, I can go that route instead, was based on the initial comment. Trying again.

Reply:
@MegaIng  Hmm you may have a point!

Reply:
(@henryiii I don't mind fixing it up myself, if you prefer)

Reply:
I think it might need to be a bound TypeVar in at least a few places, since the input matches the output. `translate` does seem to require a tree, at least by the time it starts operating on `.data` and `.children`. Translate is also missing a return annotation.

> I don't mind fixing it up myself, if you prefer

That might be best, I'll revert my change here, then you can edit the PR (or I can type ignore it, then you can follow up, whatever you like).

Reply:
No problem, fixed it in #1150.

Sorry for the mixup! It's definitely tricky code, and I wasn't the one who originally added types to it. And in fact, it caught a real edge case!